### PR TITLE
Abstract event handling from generic client

### PR DESCRIPTION
**Pull request by rranjan3**
- Move event handling logic to sdk for work order result
- Add implementation for synchronous event handling
- Add cleanup code for file I/O in ethereum listener
- Update logs to make more sense in connector

Signed-off-by: Rajeev Ranjan <rajeev2.ranjan@intel.com>

### CODE REVIEW

File: sdk/avalon_sdk/ethereum/ethereum_listener.py [hunk 1]
```diff
@@ -220,25 +229,53 @@ def __init__(self, config):
             callback(normalized_event, *kargs, **kwargs)
             self.queue.task_done()
 
-    async def start(self, eventListener, callback, *kargs, **kwargs):
+    async def sync_handler(self, *kargs, **kwargs):
```
Comment (manojgop) [inline sdk/avalon_sdk/ethereum/ethereum_listener.py]: The function name sync_handler() for an async function type looks like an oxymoron :-). I guess we could just name it like normalize_event_handler().

File: sdk/avalon_sdk/ethereum/ethereum_listener.py [hunk 2]
```diff
@@ -220,25 +229,53 @@ def __init__(self, config):
...
+        logging.info("Started synchronous handler to handle an event")
```
Comment (manojgop) [inline sdk/avalon_sdk/ethereum/ethereum_listener.py]: rephrase it as "waiting for event from queue". Also mention which function puts vents in the queue for better readability

File: sdk/avalon_sdk/ethereum/ethereum_listener.py [hunk 3]
```diff
@@ -220,25 +229,53 @@ def __init__(self, config):
             callback(normalized_event, *kargs, **kwargs)
```
Comment (manojgop) [inline sdk/avalon_sdk/ethereum/ethereum_listener.py]: Are we still using the callback with the updated event handling logic ?

File: sdk/avalon_sdk/ethereum/ethereum_listener.py [hunk 4]
```diff
@@ -220,25 +229,53 @@ def __init__(self, config):
...
-                self.listener(eventListener)) for _ in range(1)]
+        self.listeners = self.get_listeners(loop, event_filter)
         self.handlers = [loop.create_task(self.handler(
             callback, *kargs, **kwargs)) for _ in range(1)]
 
         await asyncio.gather(*self.listeners)  # infinite loop
         await self.queue.join()  # this code should never run
         await self.stop()  # this code should never run
 
+    async def get_sync_event(self, event_filter, *kargs, **kwargs):
```
Comment (manojgop) [inline sdk/avalon_sdk/ethereum/ethereum_listener.py]: get_sync_event -> get_event or something more meaningful.

File: sdk/avalon_sdk/ethereum/ethereum_listener.py [hunk 5]
```diff
@@ -220,25 +229,53 @@ def __init__(self, config):
...
+        Get a single event synchronously using the event_filter
+        provided
+        Returns an event received for the event_filter used 
+        """
+        self.queue = asyncio.Queue()
+        loop = asyncio.get_event_loop()
+        self.listeners = self.get_listeners(loop, event_filter)
+        self.handlers = [loop.create_task(self.sync_handler(*kargs, **kwargs))]
+
+        handler_result = await asyncio.gather(*self.handlers)
```
Comment (manojgop) [inline sdk/avalon_sdk/ethereum/ethereum_listener.py]: handler_result -> normalized_event or just event

File: blockchain_connector/avalon_blockchain_connector/ethereum/ethereum_connector.py [hunk 6]
```diff
@@ -238,7 +239,7 @@ def handleEvent(self, event, account, contract):
         worker_id = work_order_request["workerId"]
         requester_id = work_order_request["requesterId"]
         work_order_params = event["args"]["workOrderRequest"]
-
+        logging.info("Reveived event from blockchain")
```
[the remark this answers is not in the document]
Comment (rranjan3) [reply blockchain_connector/avalon_blockchain_connector/ethereum/ethereum_connector.py]: Addressed all.

File: sdk/avalon_sdk/ethereum/ethereum_listener.py [hunk 7]
```diff
@@ -81,15 +81,21 @@ def _is_jrpc_evt_required(self):
...
     try:
         f = open("bookmark", 'r')
         block_num = f.readline()
         return 0 if block_num == '' else int(block_num)
     except FileNotFoundError as e:
+        logging.error(e)
         return 0
+    finally:
+        if f is not None:
+            f.close()
```
[the remark this answers is not in the document]
Comment (rranjan3) [reply sdk/avalon_sdk/ethereum/ethereum_listener.py]: Updated in write. For read implementation, different handling required for error case. So cannot shrink to `with` construct.

File: sdk/avalon_sdk/ethereum/ethereum_listener.py (hunk 3 above)
Comment (rranjan3) [reply sdk/avalon_sdk/ethereum/ethereum_listener.py]: This callback is still in place for connector interaction or where multiple events are expected

**Comment by manojgop** [comment]
Added issue #321 to make the implementation similar for both fabric and ethereum. A new PR can address this issue. Most if the review comments in this PR can be handled as part of new PR

**Pull request merged**